File: app/src/main/java/openfoodfacts/github/scrachx/openfood/views/listeners/CommonBottomListener.java
package openfoodfacts.github.scrachx.openfood.views.listeners;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.res.ColorStateList;
import android.content.res.Resources;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.FragmentActivity;

import com.afollestad.materialdialogs.MaterialDialog;
import com.google.android.material.bottomnavigation.BottomNavigationView;

import openfoodfacts.github.scrachx.openfood.R;
import openfoodfacts.github.scrachx.openfood.fragments.HomeFragment;
import openfoodfacts.github.scrachx.openfood.utils.Utils;
import openfoodfacts.github.scrachx.openfood.views.ContinuousScanActivity;
import openfoodfacts.github.scrachx.openfood.views.HistoryScanActivity;
import openfoodfacts.github.scrachx.openfood.views.MainActivity;
import openfoodfacts.github.scrachx.openfood.views.ProductComparisonActivity;
import openfoodfacts.github.scrachx.openfood.views.ProductListsActivity;
import openfoodfacts.github.scrachx.openfood.views.WelcomeActivity;

public class CommonBottomListener implements BottomNavigationView.OnNavigationItemSelectedListener {
    private final Activity currentActivity;
    private final BottomNavigationView view;

    CommonBottomListener(Activity activity, BottomNavigationView view) {
        this.view = view;
        this.currentActivity = activity;
    }

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem item) {
        final Resources resources = currentActivity.getResources();
        ColorStateList color;
        switch (item.getItemId()) {
            case R.id.scan_bottom_nav:
                color = resources.getColorStateList(R.color.nav_scan);
                view.setItemIconTintList(color);
                view.setItemTextColor(color);

                openScanActivity();
                break;
            case R.id.compare_products:
                color = resources.getColorStateList(R.color.nav_compare);
                view.setItemIconTintList(color);
                view.setItemTextColor(color);

                if (isCurrentActivity(ProductComparisonActivity.class)) {
                    break;
                }
                currentActivity.startActivity((createIntent(ProductComparisonActivity.class)));
                break;
            case R.id.home_page:
            case R.id.home:
                color = resources.getColorStateList(R.color.nav_home);
                view.setItemIconTintList(color);
                view.setItemTextColor(color);

                if (isCurrentActivity(WelcomeActivity.class) || isCurrentActivity(MainActivity.class)) {
                    ((FragmentActivity) currentActivity).getSupportFragmentManager().beginTransaction().replace(R.id.fragment_container, new HomeFragment()).addToBackStack(null)
                        .commit();
                    break;
                }
                currentActivity.startActivity((createIntent(MainActivity.class)));
                break;
            case R.id.history_bottom_nav:
                color = resources.getColorStateList(R.color.nav_history);
                view.setItemIconTintList(color);
                view.setItemTextColor(color);

                if (isCurrentActivity(HistoryScanActivity.class)) {
                    break;
                }
                currentActivity.startActivity(createIntent(HistoryScanActivity.class));
                break;
            case R.id.my_lists:
                color = resources.getColorStateList(R.color.nav_lists);
                view.setItemIconTintList(color);
                view.setItemTextColor(color);

                if (isCurrentActivity(ProductListsActivity.class)) {
                    break;
                }
                currentActivity.startActivity(createIntent(ProductListsActivity.class));
                break;
            default:
        }
        return true;
    }

    private void openScanActivity() {
        // If already on the continuous scan activity, just lower the bottom sheet
        if (isCurrentActivity(ContinuousScanActivity.class)) {
            ((ContinuousScanActivity) currentActivity).collapseBottomSheet();
            return;
        }
        // If no camera is installed, alert the user
        if (!Utils.isHardwareCameraInstalled(currentActivity)) {
            new MaterialDialog.Builder(currentActivity)
                .title(R.string.no_camera_dialog_title)
                .content(R.string.no_camera_dialog_content)
                .neutralText(R.string.txtOk)
                .show();
            return;
        }
        // Otherwise check permissions and go to continuous scan activity
        if (ContextCompat.checkSelfPermission(currentActivity, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            if (currentActivity.hasWindowFocus() && ActivityCompat.shouldShowRequestPermissionRationale(currentActivity, Manifest.permission.CAMERA)) {
                new MaterialDialog.Builder(currentActivity)
                    .title(R.string.action_about)
                    .content(R.string.permission_camera)
                    .neutralText(R.string.txtOk)
                    .onNeutral((dialog, which) -> ActivityCompat
                        .requestPermissions(currentActivity, new String[]{Manifest.permission.CAMERA}, Utils.MY_PERMISSIONS_REQUEST_CAMERA))
                    .show();
            } else {
                ActivityCompat.requestPermissions(currentActivity, new String[]{Manifest.permission.CAMERA}, Utils.MY_PERMISSIONS_REQUEST_CAMERA);
            }
        } else {
            Intent intent = createIntent(ContinuousScanActivity.class);
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
            currentActivity.startActivity(intent);
        }
    }

    ////////////////////
    // Utility functions

    private boolean isCurrentActivity(Class<? extends Activity> activityClass) {
        return currentActivity != null && currentActivity.getClass().equals(activityClass);
    }

    private Intent createIntent(Class<? extends Activity> activityClass) {
        final Intent intent = new Intent(currentActivity, activityClass);
        intent.addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION);
        return intent;
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/listeners/CommonBottomListener.kt
package openfoodfacts.github.scrachx.openfood.features.listeners

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.view.MenuItem
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import androidx.fragment.app.commit
import com.afollestad.materialdialogs.MaterialDialog
import com.google.android.material.bottomnavigation.BottomNavigationView
import openfoodfacts.github.scrachx.openfood.R
import openfoodfacts.github.scrachx.openfood.features.HomeFragment
import openfoodfacts.github.scrachx.openfood.features.MainActivity
import openfoodfacts.github.scrachx.openfood.features.compare.ProductCompareActivity
import openfoodfacts.github.scrachx.openfood.features.productlists.ProductListsActivity
import openfoodfacts.github.scrachx.openfood.features.scan.ContinuousScanActivity
import openfoodfacts.github.scrachx.openfood.features.scanhistory.ScanHistoryActivity
import openfoodfacts.github.scrachx.openfood.features.welcome.WelcomeActivity
import openfoodfacts.github.scrachx.openfood.utils.MY_PERMISSIONS_REQUEST_CAMERA
import openfoodfacts.github.scrachx.openfood.utils.isHardwareCameraInstalled

class CommonBottomListener internal constructor(private val currentActivity: Activity) : BottomNavigationView.OnNavigationItemSelectedListener {
    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.scan_bottom_nav -> openScanActivity()
            R.id.compare_products -> {
                if (isCurrentActivity(ProductCompareActivity::class.java)) {
                    return true
                }
                currentActivity.startActivity(createIntent(ProductCompareActivity::class.java))
            }
            R.id.home_page, R.id.home -> {
                if (isCurrentActivity(WelcomeActivity::class.java) || isCurrentActivity(MainActivity::class.java)) {
                    (currentActivity as FragmentActivity).supportFragmentManager.commit {
                        replace(R.id.fragment_container, HomeFragment())
                        addToBackStack(null)
                    }
                    return true
                }
                currentActivity.startActivity(createIntent(MainActivity::class.java))
            }
            R.id.history_bottom_nav -> {
                if (isCurrentActivity(ScanHistoryActivity::class.java)) {
                    return true
                }
                currentActivity.startActivity(createIntent(ScanHistoryActivity::class.java))
            }
            R.id.my_lists -> {
                if (isCurrentActivity(ProductListsActivity::class.java)) {
                    return true
                }
                currentActivity.startActivity(createIntent(ProductListsActivity::class.java))
            }
            else -> return true
        }
        return true
    }

    private fun openScanActivity() {
        // If already on the continuous scan activity, just lower the bottom sheet
        if (isCurrentActivity(ContinuousScanActivity::class.java)) {
            (currentActivity as ContinuousScanActivity).collapseBottomSheet()
            return
        }
        // If no camera is installed, alert the user
        if (!isHardwareCameraInstalled(currentActivity)) {
            MaterialDialog.Builder(currentActivity).run {
                title(R.string.no_camera_dialog_title)
                content(R.string.no_camera_dialog_content)
                neutralText(R.string.txtOk)
                show()
            }

            return
        }
        // Otherwise check permissions and go to continuous scan activity
        if (ContextCompat.checkSelfPermission(currentActivity, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            if (currentActivity.hasWindowFocus() && ActivityCompat.shouldShowRequestPermissionRationale(currentActivity, Manifest.permission.CAMERA)) {
                MaterialDialog.Builder(currentActivity)
                        .title(R.string.action_about)
                        .content(R.string.permission_camera)
                        .neutralText(R.string.txtOk)
                        .onNeutral { _, _ ->
                            ActivityCompat.requestPermissions(
                                    currentActivity,
                                    arrayOf(Manifest.permission.CAMERA),
                                    MY_PERMISSIONS_REQUEST_CAMERA
                            )
                        }.show()
            } else {
                ActivityCompat.requestPermissions(currentActivity, arrayOf(Manifest.permission.CAMERA), MY_PERMISSIONS_REQUEST_CAMERA)
            }
        } else {
            val intent = createIntent(ContinuousScanActivity::class.java)
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
            currentActivity.startActivity(intent)
        }
    }

    ////////////////////
    // Utility functions
    private fun isCurrentActivity(activityClass: Class<out Activity?>) = currentActivity.javaClass == activityClass

    private fun createIntent(activityClass: Class<out Activity?>) = Intent(currentActivity, activityClass).apply {
        addFlags(Intent.FLAG_ACTIVITY_NO_ANIMATION)
    }
}